grid.v3/query.go

```diff
@@ -0,0 +1,192 @@
+package grid
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/lytics/grid/grid.v3/registry"
+)
+
+type entityType string
+
+const (
+	// Peers filter for query.
+	Peers entityType = "peer"
+	// Actors filter for query.
+	Actors entityType = "actor"
+	// Mailboxes filter for query.
+	Mailboxes entityType = "mailbox"
+)
+
+type changeType int
+
+const (
+	entityErr        changeType = 0
+	entityDiscovered changeType = 1
+	entityLost       changeType = 2
+)
+
+// QueryEvent indicating that an entity has been discovered,
+// lost, or some error has occured with the watch.
+type QueryEvent struct {
+	name   string
+	err    error
+	entity entityType
+	change changeType
+}
+
+// Name of entity that caused the event.
+func (e *QueryEvent) Name() string {
+	return e.name
+}
+
+// Discovered entity.
+func (e *QueryEvent) Discovered() bool {
+	return e.change == entityDiscovered
+}
+
+// Lost entity.
+func (e *QueryEvent) Lost() bool {
+	return e.change == entityLost
+}
+
+// Err caught watching query events. The error
+// is not associated with any particular entity.
+func (e *QueryEvent) Err() error {
+	return e.err
+}
+
+// String representation of query event.
+func (e *QueryEvent) String() string {
+	switch e.change {
+	case entityLost:
+		return fmt.Sprintf("query event: %v change: lost: %v", e.entity, e.name)
+	case entityDiscovered:
+		return fmt.Sprintf("query event: %v change: discovered: %v", e.entity, e.name)
+	default:
+		return fmt.Sprintf("query event: error: %v", e.err)
+	}
+}
+
+// QueryWatch monitors the entry and exit of peers, actors, or mailboxes.
+//
+// Example usage:
+//
+//     client, err := grid.NewClient(...)
+//     ...
+//
+//     currentpeers, watch, err := client.QueryWatch(c, grid.Peers)
+//     ...
+//
+//     for _, peer := range currentpeers {
+//         // Do work regarding peer.
+//     }
+//
+//     for event := range watch {
+//         if event.Err() != nil {
+//             // Error occured watching peers, deal with error.
+//         }
+//         if event.Lost() {
+//             // Existing peer lost, reschedule work on extant peers.
+//         }
+//         if event.Discovered() {
+//             // New peer discovered, assign work, get data, reschedule, etc.
+//         }
+//     }
+func (c *Client) QueryWatch(ctx context.Context, filter entityType) ([]string, <-chan *QueryEvent, error) {
```

Comment:
entityType (and its consts) aren't Public,  that will force the callers to redefine the consts when calling.